### DISCUSSION TITLE
Gateway health monitoring

### DISCUSSION THREAD

**Discussion by NorthBaandee** [timmbogner/Farm-Data-Relay-System#189] [Ideas]
I set up a small system with a uart gateway a solar powered lora gateway and a solar powered lora repeater and a few sensor nodes but it is difficult to troubleshoot because none of the components (except the sensor nodes themselves at the far end of the system) send signals to indicate if they are up or not. So i want to read battery volts (easy with a simple voltage divider) and maybe even solar current (with a hall sensor or something) and have the g/w's and rptrs send those readings back to base regularly. That way i can watch my dashboard and quickly see where the weak links are

**Comment by timmbogner** [comment]
Gateways have the ability to run user-defined functions at set intervals, and [it's covered here in the readme.](https://github.com/timmbogner/Farm-Data-Relay-System/blob/main/extras/Gateway.md#user-defined-functions) You set up routing for readings coming from inside the gateway by configuring ```INTERNAL_ACT```.

I should really add an example to the repo itself, though.

Sounds like a cool setup! Let me know how the internal DataReading stuff works out!

**Reply by timmbogner** [reply]
[I found and added an example that I made a long time ago.](https://github.com/timmbogner/Farm-Data-Relay-System/tree/main/examples/Gateway_Examples/Repeater_Voltage)

**Comment by NorthBaandee** [comment]
That solution looks perfect! My system is still not quite presentable enough to showcase here but this information will help. Living in this remote location means components take ages to procure, and progress is slow. I blew up a esp32 the other day through carelessness now i have to wait a month for a replacement. 8*(

**Comment by NorthBaandee** [comment]
The "toast rack" uart gateway with rpi zero node-red server. The battery is
one of these pieces of shit
https://wiki.geekworm.com/index.php/Raspi_UPS_HAT_Board
Whoever designed it has no idea how a ups is supposed to work. A few people
have hacked them but it gets messy. A WIP and currently bypassed.
Peter vk6fun

**Comment by NorthBaandee** [comment]
I still cannot properly debug anything because the serial monitor shows nothing. I have #define FDRS_DEBUG set but I see nothing. I have put Serial.begin(115200); followed by Serial.println("Here I am"); at the beginning of the setup function but nothing. I am despairing. Every other project I run initialises the serial port and I can read the messages in the serial monitor window. FDRS is the only project among the dozens I have which cannot produce any serial output whatever. It is not my cable. It is not my PC. I have tried two other PCs but nothing. Something ruins the serial port right at the beginning but what?

**Reply by timmbogner** [reply]
There's nothing more I can really offer you with your serial issues at this moment. No one else has reported problems and you haven't made it clear what boards this is occurring on. If the ESP is not displaying serial info on reset, that indicates an issue that's beyond me.

**Reply by timmbogner** [reply]
For what it's worth, you should not be using Serial.begin() or Serial.println() within FDRS. These functions are covered by beginFDRS() and DBG() respectively.

**Comment by NorthBaandee** [comment]
Usually I simply plug the thing into a USB port and I can watch the serial output on /dev/ttyUSB0. This does not happen with FDRS. I pressed the reset button. OH LOOK there's all the serial output! So simply plugging the unit into a port I get nothing but pressing the reset button makes the serial output appear. Baffling for this feeble brain. Anyway now I can see that radiolib cannot initialise my HopeRF RFM95 module with error message -2. I can make it work with this https://randomnerdtutorials.com/esp32-lora-rfm95-transceiver-arduino-ide/ but radiolib  seems to want pins this module does not have. In particular, LORA_BUSY which is set to pin 33 in the example.

**Reply by timmbogner** [reply]
You can enter ```RADIOLIB_NC```, or omit ```LORA_BUSY``` altogether, (in which case it will default to ```RADIOLIB_NC```). Check out the [LoRa Configuration](https://github.com/timmbogner/Farm-Data-Relay-System/blob/main/extras/Gateway.md#lora-configuration) section of the gateway documentation for more.

**Comment by NorthBaandee** [comment]
Commenting out LORA_BUSY wouldn't compile but setting it to RADIOLIB_NC did the job. Thank you so much!  I need to look closely at this radiolib. Hamradio protocols like APRS are particularly interesting. This protocol usually does GPS but I have made it conduct all sorts of telemetry

**Reply by timmbogner** [reply]
>Commenting out LORA_BUSY wouldn't compile

This was added at the beginning of last month, so make sure you have the latest version.

RadioLib is an incredible library, and I only use a fraction of it's capabilities.  Definitely check it out!
